### DISCUSSION TITLE
Usage of validation data for early stopping and hyperparameter tuning

### DISCUSSION THREAD

**Discussion by stinahb** [alan-turing-institute/sktime#6581] [Q&A] [answered]
Dear all,

I am working on a Human Activity Recognition task using accelerometer data and plan to utilize the HIVECOTEv2, ResNet, and InceptionTime classifiers provided by sktime. I have a few questions that I hope some of you might be able to help me with:

1. I assume that the required input data format for sktime time series classifiers is equal-length time windows of the data, represented as a 3D array with the shape (number of segments, window size, number of features). Is this assumption correct?

2. Both the ResNet and InceptionTime classifiers include a “callback” parameter. However, I am unsure how to implement “early_stopping” based on “val_loss” as a callback option, given that the classifier.fit() function does not have a “validation_data” argument. Is there a way to achieve early stopping in this context?

3. Regarding hyperparameter tuning, is it possible to use the GridSearch provided by sktime with an independent validation set for HIVECOTEv2, ResNet, and InceptionTime, or is it only feasible to perform tuning based on cross-validation?

I would greatly appreciate any assistance you can provide, even if it's only for some of the questions. Thank you very much!

Best,
Kristina

**Comment by fkiraly** [comment, marked as answer]
Welcome to `sktime`!

For a general introduction to time series classification, see here: https://www.sktime.net/en/stable/examples/02_classification.html

1. `sktime` supports multiple data container formats, some allow you to pass collections of unequal length series, such as the `pd-multiindex` specification. You can also pass as 3D `numpy` array, the convention then is (instance, variable, time) index, that is different from your ordering.

2. I believe you need to pass an `EarlyStopping` callback: https://keras.io/api/callbacks/early_stopping/ - this is not continuously tested though, so it would be appreciated if you could report on whether this works.

The estimators are also not very well documented, so improvements to the docstring would be much appreciated!

3. I am not sure if I get your question. In TSC, the instances are typically assumed independent, so would cross-validation not ensure that, in each split, the train and test fold are independent? Either way, you can pass arbitrary `sklearn` splitter objects, e.g., `GroupedKFold`, if you have groups where time series within are not independent.

**Reply by ecetsn** [reply]
Hello,

I'm looking for guidance on monitoring validation loss when using the SKTime CNN class methods. Specifically, I’m trying to find a way to fit the model and track the loss on validation data throughout the epochs. I know I can use Keras EarlyStopping, but it only monitors training loss. Is there a method to obtain the learning curve for validation loss or to prevent the model from overfitting to the training data?

Best,
Ece
